feature: Add DuckDB profile mapping

prithvijitguha opened the pull request:
## Description

<!-- Add a brief but complete description of the change. -->
- Added support for dbt-duckdb
- Added jaffle_shop repo for duckdb 
- Added example for duck db dag usage

## Related Issue(s)

<!-- If this PR closes an issue, you can use a keyword to auto-close. -->
<!-- i.e. "closes #202 " -->
- Relates to #202 
- 
## Breaking Change?

<!-- If this introduces a breaking change, specify that here. -->

## Checklist

- [x] I have made corresponding changes to the documentation (if required)
- [x] I have added tests that prove my fix is effective or that my feature works

prithvijitguha commented:
Hey all, 

It looks like more files due to the addition of the jaffle_shop_duckdb repo to the dev folder. Its the more updated one replacing the original jaffle_shop demo repo.

If you think this seems like too much for one PR, let me know, maybe I can break this up in other PRs :)

tatiana commented on cosmos/profiles/duckdb/user_pass.py:
```diff
@@ -0,0 +1,37 @@
+        profile = {**self.mapped_params, **self.profile_args}
+        return self.filter_null(profile)
+
+    @property
```
What's the advantage of writing this method in the `DuckDBUserPasswordProfileMapping` as not doing anything here and just relying on the `BaseProfileMapping` parent class implementation?

tatiana commented on dev/dags/example_duckdb_dag.py:
```diff
@@ -0,0 +1,42 @@
+basic_cosmos_dag = DbtDag(
+    # dbt/cosmos-specific parameters
+    project_config=ProjectConfig(
+        DBT_ROOT_PATH / "jaffle_shop_duckdb",
```
How different is this Jaffle shop project from the other two dbt jaffle shop projects we already have in the Cosmos repo:
- https://github.com/astronomer/astronomer-cosmos/tree/main/dev/dags/dbt/jaffle_shop
- https://github.com/astronomer/astronomer-cosmos/tree/main/dev/dags/dbt/altered_jaffle_shop

Would it be possible to use one of the existing dbt projects and use a different `profile.yml`?

prithvijitguha replied:
Yes we can! that's mainly where it differs, makes more sense to reuse what we have.

tatiana replied:
It would be really good if we did not add yet another jaffle shop copy to the Cosmos repo.
If we can reuse one of the existing ones, let's do this.
Otherwise, if we cannot for any reason, we can create a smaller dbt project for duck db, if really needed.

prithvijitguha replied:
Hi Tatiana,  we can reuse the same jaffle shop in the dev directory. I've tested with the example duckdb dag present.

tatiana reviewed (commented):
This looks fantastic! Thank you so much for your contribution, @prithvijitguha! I’ve added a bit of feedback inline for you.

prithvijitguha commented on cosmos/profiles/duckdb/user_pass.py:
```diff
@@ -0,0 +1,37 @@
+
+    airflow_connection_type: str = "duckdb"
+    dbt_profile_type: str = "duckdb"
+
```
Sure! I'll add that

prithvijitguha replied:
Hey Tatiana, 

`is_community flag` has been added.

prithvijitguha commented:
Hey @tatiana, 

Thanks for your review, let me take a closer look at the integration test failures, might be with introducing new dependency for duckdb and airflow provider. 

I'll push another commit once this is fixed.

pankajastro commented:
Hey @prithvijitguha, were you able to successfully run the Jaffle Shop project? I'm encountering a concurrency/lock-related error when running parallel tasks

```
  IO Error: Could not set lock on file "/usr/local/airflow/dags/dbt/jaffle_shop/jaffle_shop.duck_db": Conflicting lock is held in /usr/local/bin/python3.12 (PID 648). See also https://duckdb.org/docs/connect/concurrency
[2025-04-25, 18:55:51 UTC] {subprocess.py:94} INFO - Command exited with return code 2
[2025-04-25, 18:55:51 UTC] {local.py:405} INFO -   IO Error: Could not set lock on file "/usr/local/airflow/dags/dbt/jaffle_shop/jaffle_shop.duck_db": Conflicting lock is held in /usr/local/bin/python3.12 (PID 648). See also https://duckdb.org/docs/connect/concurrency
```

tatiana reviewed (approved):
Thanks a lot to @prithvijitguha for addressing the feedback and @pankajastro for fixing the integration tests in #1699. I'm approving this PR, it will temporarily break our main branch, but it should be resolved as soon as we merge Pankaj's PR.

tatiana commented on tests/profiles/duckdb/test_duckdb_user_pass.py:
```diff
@@ -0,0 +1,85 @@
+    mock_duckdb_conn: Connection,
+) -> None:
+    """
+    Tests that you can override the profile values.
```
```suggestion
    Test that users can override the profile values.
```

prithvijitguha commented:
Hey @pankajastro! 

Sorry for the late reply, yes I was able to run jaffle shop locally, I seemed to have some issues running the integration tests, I am using WSL2, I assumed that might be an issue as well.

Thank you so much for fixing the integration tests!

prithvijitguha commented:
> Thanks a lot to @prithvijitguha for addressing the feedback and @pankajastro for fixing the integration tests in #1699. I'm approving this PR, it will temporarily break our main branch, but it should be resolved as soon as we merge Pankaj's PR.

Thank you @tatiana!

Its a great experience to work on this repo. Many thanks for @pankajastro for this fix :)
